File: lc/best-time-to-buy-and-sell-stock.go
// 7 1 5 3 6 4 0 7
//   .     .
// 0 1 2 3 4 5 6 7
// 0 0 4 2 5 3 0 7
// change minbound only if you find something lower
// maxprofit is always max(maxprofit, curr - minbound)
// minbound 1
// maxprofit 4 
// corner case: with one day, prof is zero --> init maxprofit to 0
// works for mono decreasing too to give zero
// 1 5 3 -- exp 4
//     c
// minbound 1
// maxprofit 4
func maxProfit(prices []int) int {
    maxprofit, minbound := 0, prices[0]
    for _, curr := range prices {
        minbound = min(minbound, curr)
        maxprofit = max(maxprofit, curr - minbound)
    }
    return maxprofit
}

File: lc/py/best-time-to-buy-and-sell-stock.py
from typing import List
# 7 1 5 3 6 4
#   -------

# 7 3 5 1 6 4
#   ---
#   m x i <-- new min (but still ii remember my maxp)
# always update the max
# but only after you have updated the min 
# 5     4     3       8  11  1  3
#             minp 
# xp = 0
# 1 3 
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        maxp = 0
        minp = prices[0]
        for i in range(1, len(prices)):
            minp = min(prices[i], minp)
            maxp = max(maxp, prices[i] - minp)
        return maxp
        
